### PR TITLE
feat: locales package

### LINKED ISSUE #1080: Implement `lingo.dev/locale-codes` package
What We Are Building?

A JavaScript package that helps developers work with locale codes (like "en-US" or "zh-Hans-CN") and get country/language names in different languages. For example, showing "United States" as "Estados Unidos" to Spanish users.

## Why This Package Is Needed
- Apps need to show country names in the user's language
- Locale codes come in different formats (en-US vs en_US) and need parsing
- Scripts like Chinese Simplified (Hans) vs Traditional (Hant) must be handled
- Current solutions are incomplete or too heavy for modern apps

## Package Location and Publishing
- **Code Location**: `packages/locales/` in the monorepo
- **Published As**: `@lingo.dev/_locales` on npm
- **Used As**: `import { getCountryName } from "lingo.dev/locales"`
- **Same Pattern As**: `@lingo.dev/_spec` package

#### `parseLocale(locale: string): LocaleComponents`
Breaks apart a locale string into its pieces.

```javascript
parseLocale("en-US");          // { language: "en", region: "US" }
parseLocale("en_US");          // { language: "en", region: "US" }
parseLocale("zh-Hans-CN");     // { language: "zh", script: "Hans", region: "CN" }
parseLocale("zh_Hans_CN");     // { language: "zh", script: "Hans", region: "CN" }
parseLocale("es");             // { language: "es" }
parseLocale("sr-Cyrl-RS");     // { language: "sr", script: "Cyrl", region: "RS" }
```

#### `getLanguageCode(locale: string): string`
Extracts just the language part.

```javascript
getLanguageCode("en-US");      // "en"
getLanguageCode("zh-Hans-CN"); // "zh"
getLanguageCode("es-MX");      // "es"
getLanguageCode("fr_CA");      // "fr"
```

#### `getScriptCode(locale: string): string | null`
Extracts the script part (how text is written).

```javascript
getScriptCode("zh-Hans-CN");   // "Hans" (Simplified Chinese)
getScriptCode("zh-Hant-TW");   // "Hant" (Traditional Chinese)
getScriptCode("sr-Cyrl-RS");   // "Cyrl" (Cyrillic script)
getScriptCode("en-US");        // null (no script specified)
getScriptCode("es");           // null
```

#### `getRegionCode(locale: string): string | null`
Extracts the country/region part.

```javascript
getRegionCode("en-US");        // "US"
getRegionCode("zh-Hans-CN");   // "CN"
getRegionCode("es");           // null (no region specified)
getRegionCode("fr_CA");        // "CA"
```

#### `isValidLocale(locale: string): boolean`
Checks if a locale string is properly formatted and uses real codes.

```javascript
isValidLocale("en-US");        // true
isValidLocale("en_US");        // true
isValidLocale("zh-Hans-CN");   // true
isValidLocale("invalid");      // false
isValidLocale("en-FAKE");      // false
isValidLocale("xyz-US");       // false
```

#### `isValidLanguageCode(code: string): boolean`
Checks if a language code is real.

```javascript
isValidLanguageCode("en");     // true
isValidLanguageCode("zh");     // true
isValidLanguageCode("es");     // true
isValidLanguageCode("xyz");    // false
isValidLanguageCode("fake");   // false
```

#### `isValidScriptCode(code: string): boolean`
Checks if a script code is real.

```javascript
isValidScriptCode("Hans");     // true (Simplified Chinese)
isValidScriptCode("Hant");     // true (Traditional Chinese)
isValidScriptCode("Latn");     // true (Latin alphabet)
isValidScriptCode("Cyrl");     // true (Cyrillic)
isValidScriptCode("Fake");     // false
```

#### `isValidRegionCode(code: string): boolean`
Checks if a country/region code is real.

```javascript
isValidRegionCode("US");       // true
isValidRegionCode("CN");       // true
isValidRegionCode("GB");       // true
isValidRegionCode("ZZ");       // false
isValidRegionCode("FAKE");     // false
```

### 3. Name Resolution Functions (Async)

**Why These Are Async**: These functions load language-specific data files on demand. This keeps the main package small (~5KB) while supporting 200+ languages. When you ask for Spanish names, it loads the Spanish data file.

#### `getCountryName(countryCode: string, displayLanguage = "en"): Promise<string>`
Gets a country name in any language.

```javascript
// Default English
await getCountryName("US");           // "United States"
await getCountryName("CN");           // "China"
await getCountryName("DE");           // "Germany"

// Spanish
await getCountryName("US", "es");     // "Estados Unidos"
await getCountryName("CN", "es");     // "China"
await getCountryName("DE", "es");     // "Alemania"

// French
await getCountryName("US", "fr");     // "États-Unis"
await getCountryName("CN", "fr");     // "Chine"
await getCountryName("DE", "fr");     // "Allemagne"

// Chinese
await getCountryName("US", "zh");     // "美国"
await getCountryName("GB", "zh");     // "英国"
await getCountryName("JP", "zh");     // "日本"

// Arabic
await getCountryName("US", "ar");     // "الولايات المتحدة"
```

#### `getLanguageName(languageCode: string, displayLanguage = "en"): Promise<string>`
Gets a language name in any language.

```javascript
// Default English
await getLanguageName("en");          // "English"
await getLanguageName("zh");          // "Chinese"
await getLanguageName("es");          // "Spanish"

// Spanish
await getLanguageName("en", "es");    // "inglés"
await getLanguageName("zh", "es");    // "chino"
await getLanguageName("fr", "es");    // "francés"

// French
await getLanguageName("en", "fr");    // "anglais"
await getLanguageName("zh", "fr");    // "chinois"
await getLanguageName("de", "fr");    // "allemand"

// Chinese
await getLanguageName("en", "zh");    // "英语"
await getLanguageName("es", "zh");    // "西班牙语"
await getLanguageName("ar", "zh");    // "阿拉伯语"
```

#### `getScriptName(scriptCode: string, displayLanguage = "en"): Promise<string>`
Gets a script name in any language.

```javascript
// Default English
await getScriptName("Hans");          // "Simplified Han"
await getScriptName("Hant");          // "Traditional Han"
await getScriptName("Latn");          // "Latin"
await getScriptName("Cyrl");          // "Cyrillic"

// Spanish
await getScriptName("Hans", "es");    // "han simplificado"
await getScriptName("Cyrl", "es");    // "cirílico"

// Chinese
await getScriptName("Latn", "zh");    // "拉丁文"
await getScriptName("Arab", "zh");    // "阿拉伯文"
```

### Step 1: Create Package Structure
**What**: Set up folders and files.
**How**: Create `packages/locales/` with `src/`, `data/` subdirectories.
**Why**: Organization makes code maintainable.

### Step 2: Define TypeScript Types
**What**: Create data structure definitions.
**How**: Write interfaces for `LocaleComponents` and other types.
**Why**: TypeScript catches errors and improves developer experience.

### Step 3: Build Locale Parser
**What**: Function to break apart locale strings.
**How**: Use regex to split "zh-Hans-CN" into language, script, region.
**Why**: Apps get locale strings but need the individual pieces.

### Step 4: Create Validation Functions
**What**: Check if codes are real and valid.
**How**: Load official ISO code lists, check input against them.
**Why**: Prevents bugs from typos and invalid codes.

### Step 5: Get CLDR Locale Data
**What**: Download official translation data.
**How**: Get Unicode CLDR JSON files, extract country/language names.
**Why**: CLDR is the authoritative source for localized names.

### Step 6: Build Name Resolution
**What**: Functions that return localized names.
**How**: Load language-specific JSON files on demand.
**Why**: Supports many languages without bloating bundle size.

### Step 7: Configure Building
**What**: Set up compilation and packaging.
**How**: Configure TypeScript, bundling, exports.
**Why**: Create installable package for users.

### Step 8: Write Tests
**What**: Test all functions thoroughly.
**How**: Test valid/invalid inputs, multiple languages, edge cases.
**Why**: Ensure everything works correctly.

### Step 9: Add to Monorepo
**What**: Make package importable as "lingo.dev/locales".
**How**: Add export mapping in main package.json.
**Why**: Consistent with other lingo.dev packages.

### Step 10: Add Documentation
**What**: Code comments and usage docs.
**How**: JSDoc comments on all functions.
**Why**: Help developers understand usage.

## Success Criteria
- All example code above works correctly
- Package size under 10KB (core functions)
- Supports all existing display languages
- Handles both "-" and "_" delimiters
- Full TypeScript support
- Tests pass for all functions

### PR DESCRIPTION
**Pull request by leen-neel**
# Implement `lingo.dev/locales` Package

Closes #1080

## What's New

A JavaScript package for working with locale codes and getting localized country/language names. This package helps developers:

- **Parse locale strings** like "en-US" or "zh-Hans-CN" into their components
- **Validate locale codes** against real ISO standards
- **Get localized names** for countries, languages, and scripts in 200+ languages
- **Handle complex locales** with scripts (like Simplified vs Traditional Chinese)

### Locale Parsing

```typescript
parseLocale("en-US"); // { language: "en", region: "US" }
parseLocale("zh-Hans-CN"); // { language: "zh", script: "Hans", region: "CN" }
getLanguageCode("en-US"); // "en"
getScriptCode("zh-Hans-CN"); // "Hans"
getRegionCode("en-US"); // "US"
```

### Validation

```typescript
isValidLocale("en-US"); // true
isValidLocale("en-FAKE"); // false
isValidLanguageCode("en"); // true
isValidScriptCode("Hans"); // true
isValidRegionCode("US"); // true
```

### Localized Names (Async)

The name functions are asynchronous because they fetch localized data from Unicode CLDR (Common Locale Data Repository) on-demand. This approach keeps the bundle size small while providing access to 200+ languages.

```typescript
await getCountryName("US"); // "United States"
await getCountryName("US", "es"); // "Estados Unidos"
await getLanguageName("en", "fr"); // "anglais"
await getScriptName("Hans", "es"); // "han simplificado"
```

**Why Async?** The functions fetch CLDR data from GitHub raw URLs when first called for a specific language. Subsequent calls for the same language use cached data, so only the first call for each language is slow.

**Data Source:** https://github.com/unicode-org/cldr-json

## Technical Details

- **Package**: `@lingo.dev/_locales` (imported as `lingo.dev/locales`)
- **Bundle Size**: ~13KB
- **Data Source**: Unicode CLDR via GitHub raw URLs
- **Caching**: In-memory cache for network requests
- **Fallback**: English when requested language unavailable

## Tests

- 83 tests passing
- All functions covered
- Edge cases and error handling tested

## Files Added

```
packages/locales/
├── src/
│   ├── index.ts              # Main exports
│   ├── types.ts              # TypeScript interfaces
│   ├── parser.ts             # Locale parsing
│   ├── validation.ts         # Validation functions
│   ├── names/
│   │   ├── index.ts          # Name resolution
│   │   └── loader.ts         # CLDR data loader
│   └── *.spec.ts             # Tests
├── package.json
├── tsconfig.json
├── tsup.config.ts
└── README.md
```

## Usage

```typescript
import { parseLocale, getCountryName } from "lingo.dev/locales";

// Parse locales
const locale = parseLocale("zh-Hans-CN");

// Get localized names
const countryName = await getCountryName("US", "es"); // "Estados Unidos"
```

### CODE REVIEW

**Comment by VAIBHAVSING** [comment]
@leen-neel 
I think the changes are needed only in the packages/locale directory. Please remove all other changes.

**Comment by leen-neel** [comment]
> @leen-neel I think the changes are needed only in the packages/locale directory. Please remove all other changes.

done!

File: packages/locales/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,113 @@
+{
```
Comment (VAIBHAVSING) [inline packages/locales/tsconfig.json:1]: this file should be cleaner 
you can reference [_specs/tsconfig](https://github.com/lingodotdev/lingo.dev/blob/main/packages/spec/tsconfig.json)

File: packages/locales/example.mjs [hunk 2]
```diff
@@ -0,0 +1,156 @@
+import {
+  parseLocale,
```
Comment (VAIBHAVSING) [inline packages/locales/example.mjs]: this is not needed

File: packages/locales/package.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (VAIBHAVSING) [inline packages/locales/package.json:1]: update package.json according to [`_specs/package.json`](packages/spec/package.json)

**Review by VAIBHAVSING — CHANGES_REQUESTED** [review]

File: packages/locales/README.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  components: LocaleComponents;
+  delimiter: LocaleDelimiter | null;
+  isValid: boolean;
+  error?: string;
+}
+```
+
+## License
+
+ISC
```
Comment (VAIBHAVSING) [inline packages/locales/README.md]: License Apache 2.0

**Review by maxprilutskiy — APPROVED** [review]
awesome work! :)

**Review by maxprilutskiy — APPROVED** [review]

**Comment by leen-neel** [comment]
> awesome work! :)

Thank you @maxprilutskiy!
